File: btc_nn.py
#!/usr/bin/env python3
"""
BTC-USD Volatility & Price Prediction Neural Network
Uses Python with PyTorch

Input: 60 minutes of OHLCV data
Output: volatility (std dev of returns) and price at t+15
"""

import csv
import math
import torch
import torch.nn as nn
import torch.optim as optim
from torch.utils.data import DataLoader, TensorDataset

DATA_PATH = "/root/.openclaw/.openclaw/workspace/btc_usd_1m.csv"
LOOKBACK = 60      # 60 minutes of history
FORWARD = 15       # Predict 15 minutes ahead
EPOCHS = 50
BATCH_SIZE = 32
LEARNING_RATE = 0.001

# Load data
def load_data(path):
    data = []
    with open(path, 'r') as f:
        reader = csv.DictReader(f)
        for row in reader:
            data.append({
                'open': float(row['open']),
                'high': float(row['high']),
                'low': float(row['low']),
                'close': float(row['close']),
                'volume': float(row['volume'])
            })
    return data

# Normalize using min-max
def normalize(data):
    closes = [d['close'] for d in data]
    min_val = min(closes)
    max_val = max(closes)
    range_val = max_val - min_val
    
    if range_val == 0:
        range_val = 1
    
    return [{
        'open': (d['open'] - min_val) / range_val,
        'high': (d['high'] - min_val) / range_val,
        'low': (d['low'] - min_val) / range_val,
        'close': (d['close'] - min_val) / range_val,
        'volume': d['volume']  # Keep raw for now
    } for d in data], min_val, max_val

# Create features
def create_features(data, lookback, forward):
    X, y_vol, y_price = [], [], []
    
    for i in range(lookback, len(data) - forward):
        window = data[i - lookback:i]
        
        # Features: 5 values * 60 minutes = 300 features
        features = []
        for d in window:
            features.extend([d['open'], d['high'], d['low'], d['close'], d['volume']])
        
        # Targets
        forward_prices = [data[j]['close'] for j in range(i, i + forward)]
        returns = []
        for j in range(len(forward_prices) - 1):
            if forward_prices[j] != 0:
                returns.append((forward_prices[j+1] - forward_prices[j]) / forward_prices[j])
        
        volatility = math.sqrt(sum(r*r for r in returns) / len(returns)) if returns else 0.0
        forward_price = forward_prices[-1]
        
        X.append(features)
        y_vol.append(volatility)
        y_price.append(forward_price)
    
    return X, y_vol, y_price

# Model
class PredictionModel(nn.Module):
    def __init__(self, input_size):
        super().__init__()
        self.fc1 = nn.Linear(input_size, 128)
        self.fc2 = nn.Linear(128, 64)
        self.fc3 = nn.Linear(64, 32)
        self.fc4_vol = nn.Linear(32, 1)   # volatility head
        self.fc4_price = nn.Linear(32, 1) # price head
        self.relu = nn.ReLU()
        self.dropout = nn.Dropout(0.2)
    
    def forward(self, x):
        x = self.relu(self.fc1(x))
        x = self.dropout(self.relu(self.fc2(x)))
        x = self.relu(self.fc3(x))
        volatility = self.fc4_vol(x)
        price = self.fc4_price(x)
        return volatility, price

# Main
print("Loading data...")
data = load_data(DATA_PATH)
print(f"Loaded {len(data)} data points")

print("Normalizing...")
normalized, min_val, max_val = normalize(data)

print(f"Creating features (lookback: {LOOKBACK}, forward: {FORWARD})...")
X, y_vol, y_price = create_features(normalized, LOOKBACK, FORWARD)
print(f"Training samples: {len(X)}")

# Convert to tensors
X_tensor = torch.tensor(X, dtype=torch.float32)
y_vol_tensor = torch.tensor(y_vol, dtype=torch.float32).reshape(-1, 1)
y_price_tensor = torch.tensor(y_price, dtype=torch.float32).reshape(-1, 1)

# Train/val split
split = int(len(X) * 0.8)
X_train, X_val = X_tensor[:split], X_tensor[split:]
y_vol_train, y_vol_val = y_vol_tensor[:split], y_vol_tensor[split:]
y_price_train, y_price_val = y_price_tensor[:split], y_price_tensor[split:]

print(f"Training set: {len(X_train)}, Validation: {len(X_val)}")

# DataLoader
train_dataset = TensorDataset(X_train, y_vol_train, y_price_train)
train_loader = DataLoader(train_dataset, batch_size=BATCH_SIZE, shuffle=True)

# Model
model = PredictionModel(LOOKBACK * 5)
optimizer = optim.Adam(model.parameters(), lr=LEARNING_RATE)
mse = nn.MSELoss()

# Training
print("Starting training...")
for epoch in range(EPOCHS):
    model.train()
    total_loss = 0.0
    
    for batch_X, batch_y_vol, batch_y_price in train_loader:
        optimizer.zero_grad()
        
        vol_pred, price_pred = model(batch_X)
        
        loss_vol = mse(vol_pred, batch_y_vol)
        loss_price = mse(price_pred, batch_y_price)
        loss = loss_vol + loss_price
        
        loss.backward()
        optimizer.step()
        
        total_loss += loss.item()
    
    # Validation
    model.eval()
    with torch.no_grad():
        vol_pred, price_pred = model(X_val)
        val_loss_vol = mse(vol_pred, y_vol_val).item()
        val_loss_price = mse(price_pred, y_price_val).item()
    
    if (epoch + 1) % 10 == 0:
        avg_loss = total_loss / len(train_loader)
        print(f"Epoch {epoch+1}/{EPOCHS} - Train Loss: {avg_loss:.4f}, Val Vol: {val_loss_vol:.4f}, Val Price: {val_loss_price:.4f}")

print("Training complete!")

# Save model
torch.save(model.state_dict(), "/root/.openclaw/.openclaw/workspace/btc_model.pt")
print("Model saved to btc_model.pt")

File: btc_nn.rb
# frozen_string_literal: true

# BTC-USD Volatility & Price Prediction Neural Network
# Uses Ruby with Torch
#
# Input: 60 minutes of OHLCV data
# Output: volatility (std dev of returns) and price at t+15

require "torch"
require "csv"

DATA_PATH = "/root/.openclaw/.openclaw/workspace/btc_usd_1m.csv"
LOOKBACK = 60      # 60 minutes of history
FORWARD = 15       # Predict 15 minutes ahead

# Load and preprocess data
def load_data(path)
  data = []
  CSV.foreach(path, headers: true) do |row|
    data << {
      open: row["open"].to_f,
      high: row["high"].to_f,
      low: row["low"].to_f,
      close: row["close"].to_f,
      volume: row["volume"].to_f
    }
  end
  data
end

# Normalize data using min-max scaling
def normalize(data)
  closes = data.map { |d| d[:close] }
  min_val = closes.min
  max_val = closes.max
  range = max_val - min_val
  
  data.map do |d|
    {
      open: range.zero? ? 0 : (d[:open] - min_val) / range,
      high: range.zero? ? 0 : (d[:high] - min_val) / range,
      low: range.zero? ? 0 : (d[:low] - min_val) / range,
      close: range.zero? ? 0 : (d[:close] - min_val) / range,
      volume: d[:volume] # volume not normalized for now
    }
  end
end

# Create features: 60-minute lookback window
def create_features(data, lookback, forward)
  x = [] # input features
  y_vol = [] # volatility target
  y_price = [] # price target
  
  (lookback...data.length - forward).each do |i|
    window = data[i - lookback...i]
    
    # Features: open, high, low, close, volume for each minute
    features = window.flat_map { |d| [d[:open], d[:high], d[:low], d[:close], d[:volume]] }
    
    # Target: volatility (std dev of returns) and forward price
    forward_prices = data[i...i + forward].map { |d| d[:close] }
    returns = forward_prices.each_cons(2).map { |a, b| (b - a) / a }
    volatility = returns.sum.zero? ? 0.0 : Math.sqrt(returns.map { |r| r**2 }.sum / returns.size)
    forward_price = forward_prices.last
    
    x << features
    y_vol << volatility
    y_price << forward_price
  end
  
  [x, y_vol, y_price]
end

# Simple Neural Network model
class PredictionModel < Torch::NN::Module
  def initialize(input_size)
    super()
    @fc1 = Torch::NN::Linear.new(input_size, 128)
    @fc2 = Torch::NN::Linear.new(128, 64)
    @fc3 = Torch::NN::Linear.new(64, 32)
    @fc4_vol = Torch::NN::Linear.new(32, 1)   # volatility head
    @fc4_price = Torch::NN::Linear.new(32, 1) # price head
    @relu = Torch::NN::ReLU.new
    @dropout = Torch::NN::Dropout.new(0.2)
  end

  def forward(x)
    x = @relu.(@fc1.(x))
    x = @dropout.(@relu.(@fc2.(x)))
    x = @relu.(@fc3.(x))
    volatility = @fc4_vol.(x)
    price = @fc4_price.(x)
    [volatility, price]
  end
end

# Main training loop
puts "Loading data..."
data = load_data(DATA_PATH)
puts "Loaded #{data.length} data points"

puts "Normalizing..."
normalized = normalize(data)

puts "Creating features (lookback: #{LOOKBACK}, forward: #{FORWARD})..."
x, y_vol, y_price = create_features(normalized, LOOKBACK, FORWARD)
puts "Training samples: #{x.length}"

# Convert to tensors
x_tensor = Torch.tensor(x, dtype: :float)
y_vol_tensor = Torch.tensor(y_vol, dtype: :float).reshape([-1, 1])
y_price_tensor = Torch.tensor(y_price, dtype: :float).reshape([-1, 1])

# Train/val split (80/20)
split = (x.length * 0.8).to_i
x_train, x_val = x_tensor[0...split], x_tensor[split..-1]
y_vol_train, y_vol_val = y_vol_tensor[0...split], y_vol_tensor[split..-1]
y_price_train, y_price_val = y_price_tensor[0...split], y_price_tensor[split..-1]

puts "Training set: #{x_train.shape[0]}, Validation: #{x_val.shape[0]}"

# Model
model = PredictionModel.new(LOOKBACK * 5)
optimizer = Torch::ADAM.new(model.parameters, lr: 0.001)
mse = Torch::NN::MSELoss.new

# Training
epochs = 50
batch_size = 32

puts "Starting training..."
epochs.times do |epoch|
  model.train
  
  # Mini-batch training
  total_loss = 0.0
  num_batches = 0
  
  x_train.each_slice(batch_size).each do |batch|
    optimizer.zero_grad
    
    vol_pred, price_pred = model.forward(batch)
    
    loss_vol = mse.call(vol_pred, y_vol_train[0...vol_pred.shape[0]])
    loss_price = mse.call(price_pred, y_price_train[0...price_pred.shape[0]])
    loss = loss_vol + loss_price
    
    loss.backward
    optimizer.step
    
    total_loss += loss.item
    num_batches += 1
  end
  
  # Validation
  model.eval
  val_loss_vol = 0.0
  val_loss_price = 0.0
  Torch.no_grad do
    vol_pred, price_pred = model.forward(x_val)
    val_loss_vol = mse.call(vol_pred, y_vol_val).item
    val_loss_price = mse.call(price_pred, y_price_val).item
  end
  
  if (epoch + 1) % 10 == 0
    puts format("Epoch %d/%d - Train Loss: %.4f, Val Vol: %.4f, Val Price: %.4f", 
                epoch + 1, epochs, total_loss / num_batches, val_loss_vol, val_loss_price)
  end
end

puts "Training complete!"

# Save model
Torch.save(model.state_dict, "/root/.openclaw/.openclaw/workspace/btc_model.pt")
puts "Model saved to btc_model.pt"